Add a lib.vim file with some basic functionality.

artasparks opened the pull request:
This is a relatively small commit to get the conversation/work going.
I'm especially curious about your thoughts about testing. I think it
might make sense to require a real bazel instal to do real-integration
tests. Faking out the world will get a lot less realistic test coverage.

From/inspired by piperlib and blazelib.

artasparks committed:
Add a lib.vim file with some basic functionality.

This is a relatively small commit to get the conversation/work going.
I'm especially curious about your thoughts about testing. I think it
might make sense to require a real bazel instal to do real-integration
tests. Faking out the world will get a lot less realistic test coverage.

From/insipired by piperlib and blazelib.

artasparks commented:
ping =)

dbarnett commented on autoload/bazel/lib.vim:
```diff
@@ -0,0 +1,52 @@
+  if !isdirectory(l:path)
+    let l:path = fnamemodify(l:path, ':h')
+  endif
+  return maktaba#path#Join([l:path, ''])
```
`maktaba#path#AsDir(l:path)` should do the same but a little more self-documenting (and cheaper).

artasparks replied:
Done

dbarnett commented on autoload/bazel/lib.vim:
```diff
@@ -0,0 +1,52 @@
+  let l:components = maktaba#path#Split(path)
+  let l:working = ''
+  let l:workspace_path = ''
+  for l:comp in l:components
```
Would it work to use `findfile()` instead of a custom traversal loop in vimscript?

artasparks replied:
Ah, forgot about findfile. I'll do that.

dbarnett reviewed (changes requested):
We'll want to create a separate repo for the lib so that other plugins could use the utilities too, won't we?

artasparks commented:
> We'll want to create a separate repo for the lib so that other plugins could use the utilities too, won't we?

I think we should wait until the other plugins come along. For, now I'd say YAGNI. Dependencies in Vim are not super nice and come with a cost. So, if there isn't a compelling reason to do so now, I'd say we wait.

FWIW, Some small functionality could be copied if need be or refactored into a library. I can rename lib.vim to repo.vim to be clear.

dbarnett commented:
Okay, yeah, after talking about it w/ Josh a little more, I'd suggest inlining the code until we actually have a need to share it with another plugin, and just not calling it a "library" everywhere since it's not one in its current form. I'd suggest `bazel#workspace#` since that (and I'd expect any other bazel utilities) would be operating at the level of a "workspace" instead of a "repo". Other than that, I'm fine with the code itself besides those 2 comments I sent earlier.

Re your testing question:
> I'm especially curious about your thoughts about testing. I think it might make sense to require a real bazel instal to do real-integration tests.

I guess it depends. I really like having vroom tests as self-contained as possible and I expect for most things we could just use autoload functions as needed to give us higher-level override points so we could cover it in vroom at the level that adds the most value wherever faking out system calls gets ugly. Making them "integration tests" that depend on real bazel would add some new complexity challenges, and I also like that vroom tests tend to capture the signature of the tools they integrate with so when the behavior changes it's easy to compare and understand why.

artasparks committed:
Update PR with changes

- Move lib to workspace file.
- Make a simple fixture directory
- Add some workspace lib tests.

artasparks commented on vroom/bazel.vroom:
```diff
@@ -9,7 +9,6 @@ The main command this plugin provides is :Bazel, which drops to a shell and
 prompt until explicitly dismissed.
 
-  :Bazel info
+  :Bazel info <CR>
```
For some reason,  vroom always just hung unless I added this. I couldn't get the input functionality to recognize `> <CR>` with neovim. This seems to work though.

artasparks replied:
And by hang, I mean it would pause waiting for ENTER.

artasparks committed:
Fix semantics to return the dir rather than the file.

dbarnett commented on autoload/bazel/workspace.vim:
```diff
@@ -0,0 +1,51 @@
+"
+" I.e., if you are in directory: /Users/$USER/myproject/foo/bar/biff
+" And there is a WORKSPACE file at: /Users/$USER/myproject/WORKSPACE
+" Return /home/$USER/myproject. If no workspace can be found, return ''
```
Could this throw a NotFound error instead of returning `''`? A caller could always catch and ignore errors if desired.

artasparks replied:
I could; this logic came from internal code, so I'm inclined to leave it.. It's not clear to me that throwing an exception is better; having written so much go code recently, I would towards returning a raw type because it leads to explicit handling.

dbarnett replied:
In go the explicit handling is usually nil or some special type that's not likely to accidentally be used as a real value, right? My concern here is that dummy paths would in a lot of cases seem to work but do the wrong thing (and vimmers IME pretty much never handle corner cases until they blow up in practice). And IIRC the pre-existing code just inherited that from vim's path handling helpers before we started noticing what gnarly bugs we got thanks to that behavior, for instance fnamemodify silently expanding it to CWD.

artasparks replied:
Done

dbarnett commented on autoload/bazel/workspace.vim:
```diff
@@ -0,0 +1,51 @@
+" @private
+" Returns the root path for a Bazel workspace file (given the current working
+" directory). {path} is the absolute path.
+function! bazel#workspace#GetBazelRootForCwd() abort
```
Could callers just pass the `getcwd()` to GetBazelRootPath themselves? Seems better for separation of concerns and having less "surface area" to library code.

artasparks replied:
Sure, removed.

dbarnett reviewed (commented):
I forgot the context on this. Can you refresh my memory what was the use case you had in mind for it?

artasparks commented:
Context is wading into blaze.vim stuff by porting some non controversial methods =)

artasparks committed:
Remove CWD method.

artasparks committed:
Some cleanups and fixes

- Rename fixture to workspace fixture
- throw an error when searching for workspace fails

artasparks committed:
Add a test for empty path and change the exception type

dbarnett commented on autoload/bazel/workspace.vim:
```diff
@@ -15,7 +15,7 @@
   let l:path = s:AbsoluteDirectoryPath(a:path)
   if empty(l:path)
-    throw maktaba#error#NotFound('No workspace found')
+    throw maktaba#error#InvalidArguments('Empty path')
```
I think BadValue fits better, and `:h ERROR(InvalidArguments)` says to prefer BadValue if relevant.

artasparks committed:
Change IllegalArgument to BadValue
